WIP: GUI refactoring & improvements

Florastamine opened the pull request:
There was a PR here. It's gone now.

Commit:
Change the code to read the test list from an external file

Commit:
Don't hardcode testlist.txt

Commit:
Revert "Don't hardcode testlist.txt"

This reverts commit a8ccadbf87a4ef0d3c2686582ee64a84a53cf54d.

Commit:
Minor refactoring + allow Targa formats to load as normal

Commit:
Fixed JPG writer

Commit:
Write grayscale JPG

Commit:
Use video::convertColor() for the JPEG writer

Commit:
Convert EF_R8G8B8_UNORM

Commit:
Fixed JPG writer, fixed warning + bump the default quality a bit

Commit:
Replaced color converters for TGA loaders + flipping; dropped 8bit nongrayscale

Commit:
Fixed the TGA writers

Commit:
Fixed PNG loader & writer

Commit:
Fixed texture flipping for JPG

Commit:
Fixed flipping for PNG

Commit:
Fixed flipping for TGA + fixed incorrect stride

Commit:
Read/write TGA gamma data

Commit:
Parentheses

Commit:
TGA related fixes

Commit:
Don't do save operations on unwriteable formats (DDS, BMP); allows notation on the testfile

Commit:
DDS: Refactoring this mess + removed dead code

Commit:
Switched to use EF_R8_SRGB

Commit:
NULLPTR EVERYTHANG

Commit:
Testing commit

Commit:
Dropped Exists()

devshgraphicsprogramming committed:
Update Readme.md

draft, needs a refactor

Commit:
Dropped ssprintf()

manhnt9 commented on examples_tests/13.BRDF_Explorer/BRDFExplorerApp.cpp:
```diff
@@ -506,29 +506,35 @@ void BRDFExplorerApp::eventAOTextureBrowse(const ::CEGUI::EventArgs&)
         const auto image = ext::cegui::loadImage(p.second.c_str());
         loadTextureSlot(ETEXTURE_SLOT::TEXTURE_AO, image.buffer, image.w, image.h);
 
+        char buffer[1024];
```
Let's use tooltip as variable name instead of buffer like this change: https://github.com/manhnt9/IrrlichtBAW/pull/1/files/1c2ff50f43ecc04c8a51dd00a738eb636a4a71be..9ea3806325ca80e265b18185e0da7eea4af3e0b6#diff-060531d0748dffb86f3279465a244547R627

Florastamine replied:
Ouch... nice catch.

manhnt9 commented on examples_tests/13.BRDF_Explorer/BRDFExplorerApp.cpp:
```diff
@@ -506,29 +506,35 @@ void BRDFExplorerApp::eventAOTextureBrowse(const ::CEGUI::EventArgs&)
-
-    if (ext::cegui::Exists(box->getText().c_str())) {
+    
+    auto fs = device->getFileSystem();
```
Where does `device` come from?

Florastamine replied:
Hmmmmmm... Now that I've thought about it, you'd have to have a `IrrlichtDevice*` available to use Irrlicht's filesystem-related utilities. And that's not nice considering `device` lives at `main()`. Suggestions?

Florastamine replied:
IMO helpers should be a standalone piece of function as-is, not having to depend on `IrrlichtDevice*` or anything else. But it's you guys' decision.

devshgraphicsprogramming replied:
BRDFExplorerApp can keep an IrrlichtDevice around

manhnt9 replied:
Yeah, add new member to the app, you can refer to my latest commit https://github.com/buildaworldnet/IrrlichtBAW/pull/291/commits/d51bb8e8c3cf201fc90c4b17ed7ea27348d8c37f

manhnt9 replied:
Remember the forward declaration (use it as much as possible).

manhnt9 replied:
@Florastamine do you still have questions regarding this?

Florastamine replied:
It's all good

devshgraphicsprogramming commented on examples_tests/13.BRDF_Explorer/BRDFExplorerApp.cpp:
```diff
@@ -506,29 +506,35 @@ void BRDFExplorerApp::eventAOTextureBrowse(const ::CEGUI::EventArgs&)
         loadTextureSlot(ETEXTURE_SLOT::TEXTURE_AO, image.buffer, image.w, image.h);
 
+        char buffer[1024];
+        sprintf(buffer, "%s (%ix%i)\nLeft-click to select a new texture.",
```
sprintf is really unportable between compilers, use ostringstream

Florastamine replied:
There, better?

Commit:
sprintf -> std::ostringstream

Commit:
Fix

devshgraphicsprogramming committed:
Split the rules

devshgraphicsprogramming committed:
Merge pull request #300 from buildaworldnet/extension_rules

Updated Extension Rules

ghost commented:
I'll keep `Replace()` & change it to use `CEGUI::String`, since their class doesn't have a way to find-and-replace at the same time - which is what we need.

manhnt9 commented:
Note: I'll squash when merging since it has trailing test commit and overall the commits are very small.

Commit:
Proper format conversion for EF_R8_UNORM

Commit:
Added testing suite

Commit:
Added 32-bit RLE TGAs

Commit:
Correct handling of TGA header parsing

Commit:
Replaced 8-bit TGA test image with the one from npedotnet

Commit:
Fixed Linux build errors

ghost commented:
OK, also I can see you haven't ported the public GUI APIs so the second item will be deferred for now

devshgraphicsprogramming committed:
Update Ext Readme

About a nice CMake script for the sRGB branch that can define an irr::ext:: library for any extension

manhnt9 commented:
There are conflicts because I've renamed CEGUI.h/.cpp to ExtCEGUI.h/.cpp. You can re-do it in your commit too.

Commit:
Bails out with error message on unknown colorspace

Commit:
Merge branch 'master' of https://github.com/buildaworldnet/IrrlichtBAW into sRGB

Commit:
Helpers to convert linear -> sRGB

manhnt9 commented:
I think you can update this PR and we'll merge, it's good enough to be integrated into my branch now before it gets outdated for some reason may come. Other functions that are connected to the rendering back-end will probably be ported next month since the Explorer is waiting for sRGB and GLSL functions. In general, I also prefer early integration to prevent long leftover of things.

Commit:
Fixed incorrect line reading on certain platforms where CR/LF bytes are inconsistent.

Commit:
Missing #include

Commit:
SColorf::fromSRGB(), SColorf::toSRGB()

Commit:
sad violin intensifies

Commit:
Merge branch 'sRGB' of https://github.com/manhnt9/IrrlichtBAW into GUI-refactor

Commit:
Merge branch 'sRGB' of https://github.com/manhnt9/IrrlichtBAW into GUI-refactor

Commit:
Merge branch 'sRGB' of https://github.com/Florastamine/IrrlichtBAW into GUI-refactor

Commit:
Merge branch 'GUI-refactor' of https://github.com/Florastamine/IrrlichtBAW into GUI-refactor

ghost commented:
Opening a new PR. Your most recent commit also inteferes with the files I'm working on (which I already committed), and now somehow I couldn't managed to merge upstream anymore.

manhnt9 commented:
I instructed above that you'd just need to rename 2 files.
